### ISSUE TITLE
RFC: Delayed Chief governance

### ISSUE THREAD

**Issue by asymmetric** [makerdao/dss-deploy#22]
## Summary

Add a DSPause between the hat and the Chief.

## Motivation

Currently, the `hat` can effect instantaneous changes on the Chief, because the Chief is [its own `authority`](https://etherscan.io/address/0x8e2a84d6ade1e7fffee039a35ef5f19f13057152#readContract).

This opens up governance to the risk that a malicious spell could, for example, authenticate arbitrary addresses against the core (via the Chief's capabilities infrastructure).

Adding a `DSPause` in front of the Chief allows governance to catch malicious plans before they are enacted, and take countermeasures.

## Detailed design

This is the current design:

![chief before](https://user-images.githubusercontent.com/101816/56734241-cfedf080-6751-11e9-8e1e-65f6dc356fea.jpg)

This RFC proposes modifying the current DSChief to [permit mutation](https://github.com/dapphub/ds-chief/blob/58a02ff50bdbf5729252aab91bc468a3e77be6bb/src/chief.sol#L156-L164) of its `owner` and `authority` fields, to achieve the following design:

![chief after(1)](https://user-images.githubusercontent.com/101816/56742125-b05fc380-6763-11e9-8e39-18819587cd45.jpg)

### Implementation

```
Chief = new Chief(owner: 0x0, authority: self)
pause = new Pause(owner: 0x0, authority: Chief)
hat = new Spell("Chief.setOwner(pause) && Chief.setAuthority(0x0)")
// ...voting voting voting...
Chief.lift(hat)
hat.exec()
```

## Drawbacks

This design requires a modiication and a redeploy of the curent Chief. Depending on the migration path we have in mind, this could be complicated to pull off.

## Alternatives

The current Chief could be left in place, and other measures added to evaluate spells and make sure they don't call into the Chief (or only do so in allowed ways).

## Unresolved questions

Migration from the current Chief has not been explored, as it's not clear whether it would take place or if instead we'll deploy MCD auth'ed against a completely new Chief anyway.

**Comment by xwvvvvwx** [comment]
Nice catch! :tada:. As we discussed today, an attacker could use this to:

- Authorize a malicious actor on the pause, allowing them to plot plans without having to pass an MKR vote
- Bypass the delay on MKR minting (as proposed in: https://github.com/makerdao/dss-deploy/issues/21)

In the first case, the pause's delay should give honest MKR holders time to either de-authorize the malicious actor or trigger emergency shutdown, so attackers would still have to bypass additional security layers before being able to damage the system.

In the second case however, there are no additional defense mechanisms, and once the malicious proposal has become the hat, it could immediately begin to mint new MKR.

The question that I don't know the answer to right now, is whether it's worth the cost (and danger) of a chief upgrade to mitigate this attack. There are obvious difficulties in coordinating MKR holders, and we risk having MKR split between the old and new voting contracts, meaning that neither of them has enough MKR to effectively secure the system.

**Comment by gbalabasquer** [comment]
I get the `hat` can change things in the `chief` itself. But do not understand what you mean here: `This opens up governance to the risk that a malicious spell could, for example, authenticate arbitrary addresses against the core (via the Chief's capabilities infrastructure).`
The only way to access to the core contracts now is via the pause. So even if you take over the `chief`, you need to pass the delay in the pause to authorize a new contract in the core ones.
However taking over the chief doesn't sound good either. So as you say an elected `hat` could change the ownership and authority of the `chief`, which is a problem, but I guess in that case there is no other way than emergency shutdown. I'm not sure what would be the difference of having the pause before the chief.

**Comment by gbalabasquer** [comment]
Reading #21 I get the pause being before the chief would block changing MKR authored functions without the delay. Then it would make sense.

**Comment by xwvvvvwx** [comment]
@gbalabasquer 

> an elected `hat` could change the ownership and authority of the `chief`

Just a lil nitpick: `setOwner` and `setAuthority` are overridden on the chief and will always revert (see [here](https://github.com/dapphub/ds-chief/blob/master/src/chief.sol#L156)), so the `hat` cannot change ownership or authority. The `hat` can however manipulate roles to give an external party direct access to anything that has the chief as it's authority.

**Comment by gbalabasquer** [comment]
@xwvvvvwx right. That is why we shouldn't use the roles of the chief at all and instead use other contracts after the pause for those purposes (roles, guards, etc). I'd actually remove the roles thing from the chief and just do a `can_call` function that allows to execute things from the `hat` (if we are doing some rewriting of it).

**Comment by kmbarry1** [comment]
Can we view this as closed? There seems to be rough consensus on @gbalabasquer's proposal above: the Chief shouldn't be used as the authority for anything other than the Pauses that protect the rest of the system, and a set of post-pause DSAuthorities can handle fine-grained authorization (these authorities are presumably owned by appropriate pauses).

**Comment by Shrey33** [comment]
@asymmetric can you tell me how to reproduce the issue ?

**Comment by gbalabasquer** [comment]
> @asymmetric can you tell me how to reproduce the issue ?

do you really think that you can upload this as an issue in hackerOne? Please don't lose your time.
